File: src/agent_search_1688/credentials.py
"""供应商凭证存取；秘密永远不进入普通 config.json。"""

from __future__ import annotations

import json
import os
from pathlib import Path
import platform
import shutil
import stat
import subprocess

from .config import get_1688_purchase_home


OPENAI_API_KEY_ENV = "OPENAI_API_KEY"
KEYCHAIN_SERVICE = "as1688.openai.api-key"
KEYCHAIN_ACCOUNT = "openai"


class PurchaseCredentialError(RuntimeError):
    pass


def _credential_file_path() -> Path:
    return get_1688_purchase_home() / "credentials.json"


def _is_1688_keychain_item_not_found(
    result: subprocess.CompletedProcess[str],
) -> bool:
    error_text = result.stderr.lower()
    return result.returncode == 44 or "could not be found" in error_text


def _delete_1688_credential_file_if_present() -> bool:
    path = _credential_file_path()
    if not (path.exists() or path.is_symlink()):
        return False
    try:
        path.unlink()
    except OSError as exc:
        raise PurchaseCredentialError(f"无法删除凭证文件：{path}") from exc
    return True


def _load_1688_keychain_api_key() -> str | None:
    if platform.system() != "Darwin":
        return None
    security_path = shutil.which("security")
    if security_path is None:
        return None
    try:
        result = subprocess.run(
            [
                security_path,
                "find-generic-password",
                "-a",
                KEYCHAIN_ACCOUNT,
                "-s",
                KEYCHAIN_SERVICE,
                "-w",
            ],
            capture_output=True,
            text=True,
            timeout=15,
            check=False,
        )
    except (OSError, subprocess.TimeoutExpired):
        return None
    value = result.stdout.strip()
    return value if result.returncode == 0 and value else None


def _load_1688_credential_file_api_key() -> str | None:
    path = _credential_file_path()
    if not path.exists():
        return None
    try:
        if path.is_symlink():
            raise PurchaseCredentialError("凭证文件不能是符号链接")
        if stat.S_IMODE(path.stat().st_mode) & 0o077:
            raise PurchaseCredentialError(
                f"凭证文件权限过宽，请执行：chmod 600 {path}"
            )
        payload = json.loads(path.read_text(encoding="utf-8"))
    except (OSError, json.JSONDecodeError) as exc:
        raise PurchaseCredentialError(f"凭证文件损坏：{path}") from exc
    if not isinstance(payload, dict):
        raise PurchaseCredentialError("凭证文件格式无效")
    value = payload.get("openai_api_key")
    if value is None:
        return None
    if not isinstance(value, str) or not value.strip():
        raise PurchaseCredentialError("OpenAI API Key 格式无效")
    return value.strip()


def load_1688_openai_api_key(
    environ: dict[str, str] | None = None,
) -> tuple[str | None, str]:
    environment = environ if environ is not None else os.environ
    environment_key = environment.get(OPENAI_API_KEY_ENV, "").strip()
    if environment_key:
        return environment_key, f"environment:{OPENAI_API_KEY_ENV}"

    keychain_key = _load_1688_keychain_api_key()
    if keychain_key:
        return keychain_key, "macos-keychain"

    file_key = _load_1688_credential_file_api_key()
    if file_key:
        return file_key, "credential-file"
    return None, "not-configured"


def save_1688_openai_api_key(api_key: str) -> str:
    value = api_key.strip()
    if not value or any(character.isspace() for character in value):
        raise PurchaseCredentialError("OpenAI API Key 不能为空或包含空白字符")

    if platform.system() == "Darwin":
        security_path = shutil.which("security")
        if security_path is not None:
            try:
                result = subprocess.run(
                    [
                        security_path,
                        "add-generic-password",
                        "-U",
                        "-a",
                        KEYCHAIN_ACCOUNT,
                        "-s",
                        KEYCHAIN_SERVICE,
                        "-w",
                    ],
                    input=value + "\n",
                    capture_output=True,
                    text=True,
                    timeout=15,
                    check=False,
                )
            except (OSError, subprocess.TimeoutExpired) as exc:
                raise PurchaseCredentialError(
                    "无法访问 macOS 钥匙串，OpenAI API Key 没有修改"
                ) from exc
            if result.returncode != 0:
                raise PurchaseCredentialError(
                    "macOS 钥匙串拒绝保存，OpenAI API Key 没有修改"
                )
            if _load_1688_keychain_api_key() != value:
                raise PurchaseCredentialError(
                    "macOS 钥匙串保存后校验失败"
                )
            _delete_1688_credential_file_if_present()
            return "macos-keychain"

    path = _credential_file_path()
    path.parent.mkdir(mode=0o700, parents=True, exist_ok=True)
    temporary_path = path.with_suffix(".tmp")
    temporary_path.write_text(
        json.dumps({"openai_api_key": value}, ensure_ascii=False) + "\n",
        encoding="utf-8",
    )
    temporary_path.chmod(0o600)
    temporary_path.replace(path)
    path.chmod(0o600)
    if _load_1688_credential_file_api_key() != value:
        raise PurchaseCredentialError("凭证文件保存后校验失败")
    return "credential-file"


def delete_1688_openai_api_key() -> list[str]:
    """删除 as1688 自己保存的 OpenAI Key，不修改环境变量。"""

    removed_sources: list[str] = []
    if platform.system() == "Darwin":
        security_path = shutil.which("security")
        if security_path is not None:
            try:
                result = subprocess.run(
                    [
                        security_path,
                        "delete-generic-password",
                        "-a",
                        KEYCHAIN_ACCOUNT,
                        "-s",
                        KEYCHAIN_SERVICE,
                    ],
                    capture_output=True,
                    text=True,
                    timeout=15,
                    check=False,
                )
            except (OSError, subprocess.TimeoutExpired) as exc:
                raise PurchaseCredentialError("无法访问 macOS 钥匙串") from exc
            if result.returncode == 0:
                removed_sources.append("macos-keychain")
            elif not _is_1688_keychain_item_not_found(result):
                raise PurchaseCredentialError(
                    "macOS 钥匙串拒绝删除 OpenAI API Key"
                )

    if _delete_1688_credential_file_if_present():
        removed_sources.append("credential-file")
    return removed_sources

File: src/credentials.ts
import {
  chmodSync,
  existsSync,
  lstatSync,
  mkdirSync,
  readFileSync,
  renameSync,
  rmSync,
  writeFileSync,
} from "node:fs";
import { delimiter, dirname, join } from "node:path";
import { spawnSync } from "node:child_process";

import { getPurchaseHome } from "./config.js";

export const OPENAI_API_KEY_ENV = "OPENAI_API_KEY";
export const KEYCHAIN_SERVICE = "as1688.openai.api-key";
export const KEYCHAIN_ACCOUNT = "openai";

export class PurchaseCredentialError extends Error {
  override readonly name = "PurchaseCredentialError";
}

export interface CredentialOptions {
  readonly appHome?: string;
  readonly environ?: Readonly<Record<string, string | undefined>>;
  readonly platform?: NodeJS.Platform;
  readonly securityPath?: string;
}

export interface LoadedCredential {
  readonly apiKey?: string;
  readonly source: string;
}

export function loadOpenAiApiKey(options: CredentialOptions = {}): LoadedCredential {
  const environ = options.environ ?? process.env;
  const environmentKey = environ[OPENAI_API_KEY_ENV]?.trim();
  if (environmentKey) {
    return { apiKey: environmentKey, source: `environment:${OPENAI_API_KEY_ENV}` };
  }
  const keychainKey = loadKeychainApiKey(options);
  if (keychainKey) return { apiKey: keychainKey, source: "macos-keychain" };
  const fileKey = loadCredentialFileApiKey(options);
  if (fileKey) return { apiKey: fileKey, source: "credential-file" };
  return { source: "not-configured" };
}

export function saveOpenAiApiKey(apiKey: string, options: CredentialOptions = {}): string {
  const value = apiKey.trim();
  if (!value || /\s/u.test(value)) {
    throw new PurchaseCredentialError("OpenAI API Key 不能为空或包含空白字符");
  }
  const securityPath = resolveSecurityPath(options);
  if ((options.platform ?? process.platform) === "darwin" && securityPath) {
    const result = spawnSync(securityPath, [
      "add-generic-password", "-U", "-a", KEYCHAIN_ACCOUNT, "-s", KEYCHAIN_SERVICE, "-w",
    ], {
      input: `${value}\n`,
      encoding: "utf8",
      timeout: 15_000,
      stdio: ["pipe", "pipe", "pipe"],
    });
    if (result.error) {
      throw new PurchaseCredentialError("无法访问 macOS 钥匙串，OpenAI API Key 没有修改", { cause: result.error });
    }
    if (result.status !== 0) {
      throw new PurchaseCredentialError("macOS 钥匙串拒绝保存，OpenAI API Key 没有修改");
    }
    if (loadKeychainApiKey({ ...options, securityPath }) !== value) {
      throw new PurchaseCredentialError("macOS 钥匙串保存后校验失败");
    }
    deleteCredentialFileIfPresent(options);
    return "macos-keychain";
  }

  const path = credentialFilePath(options);
  mkdirSync(options.appHome ?? getPurchaseHome(pathOptions(options)), { recursive: true, mode: 0o700 });
  const temporaryPath = join(dirname(path), ".credentials.tmp");
  writeFileSync(temporaryPath, `${JSON.stringify({ openai_api_key: value })}\n`, { encoding: "utf8", mode: 0o600 });
  chmodSync(temporaryPath, 0o600);
  renameSync(temporaryPath, path);
  chmodSync(path, 0o600);
  if (loadCredentialFileApiKey(options) !== value) {
    throw new PurchaseCredentialError("凭证文件保存后校验失败");
  }
  return "credential-file";
}

export function deleteOpenAiApiKey(options: CredentialOptions = {}): string[] {
  const removed: string[] = [];
  const securityPath = resolveSecurityPath(options);
  if ((options.platform ?? process.platform) === "darwin" && securityPath) {
    const result = spawnSync(securityPath, [
      "delete-generic-password", "-a", KEYCHAIN_ACCOUNT, "-s", KEYCHAIN_SERVICE,
    ], { encoding: "utf8", timeout: 15_000, stdio: ["ignore", "pipe", "pipe"] });
    if (result.error) throw new PurchaseCredentialError("无法访问 macOS 钥匙串", { cause: result.error });
    if (result.status === 0) {
      removed.push("macos-keychain");
    } else if (!(result.status === 44 || result.stderr.toLowerCase().includes("could not be found"))) {
      throw new PurchaseCredentialError("macOS 钥匙串拒绝删除 OpenAI API Key");
    }
  }
  if (deleteCredentialFileIfPresent(options)) removed.push("credential-file");
  return removed;
}

function credentialFilePath(options: CredentialOptions): string {
  return join(options.appHome ?? getPurchaseHome(pathOptions(options)), "credentials.json");
}

function loadCredentialFileApiKey(options: CredentialOptions): string | undefined {
  const path = credentialFilePath(options);
  if (!existsSync(path)) return undefined;
  let payload: unknown;
  try {
    const stat = lstatSync(path);
    if (stat.isSymbolicLink()) throw new PurchaseCredentialError("凭证文件不能是符号链接");
    if ((stat.mode & 0o077) !== 0) {
      throw new PurchaseCredentialError(`凭证文件权限过宽，请执行：chmod 600 ${path}`);
    }
    payload = JSON.parse(readFileSync(path, "utf8"));
  } catch (error) {
    if (error instanceof PurchaseCredentialError) throw error;
    throw new PurchaseCredentialError(`凭证文件损坏：${path}`, { cause: error });
  }
  if (typeof payload !== "object" || payload === null || Array.isArray(payload)) {
    throw new PurchaseCredentialError("凭证文件格式无效");
  }
  const value = (payload as Record<string, unknown>).openai_api_key;
  if (value === undefined) return undefined;
  if (typeof value !== "string" || !value.trim()) {
    throw new PurchaseCredentialError("OpenAI API Key 格式无效");
  }
  return value.trim();
}

function loadKeychainApiKey(options: CredentialOptions): string | undefined {
  if ((options.platform ?? process.platform) !== "darwin") return undefined;
  const securityPath = resolveSecurityPath(options);
  if (!securityPath) return undefined;
  const result = spawnSync(securityPath, [
    "find-generic-password", "-a", KEYCHAIN_ACCOUNT, "-s", KEYCHAIN_SERVICE, "-w",
  ], { encoding: "utf8", timeout: 15_000, stdio: ["ignore", "pipe", "pipe"] });
  if (result.error || result.status !== 0) return undefined;
  return result.stdout.trim() || undefined;
}

function deleteCredentialFileIfPresent(options: CredentialOptions): boolean {
  const path = credentialFilePath(options);
  if (!existsSync(path) && !safeIsSymlink(path)) return false;
  try {
    rmSync(path);
    return true;
  } catch (error) {
    throw new PurchaseCredentialError(`无法删除凭证文件：${path}`, { cause: error });
  }
}

function safeIsSymlink(path: string): boolean {
  try {
    return lstatSync(path).isSymbolicLink();
  } catch {
    return false;
  }
}

function resolveSecurityPath(options: CredentialOptions): string | undefined {
  if (options.securityPath !== undefined) return options.securityPath || undefined;
  if ((options.platform ?? process.platform) !== "darwin") return undefined;
  if (existsSync("/usr/bin/security")) return "/usr/bin/security";
  const pathValue = (options.environ ?? process.env).PATH ?? "";
  for (const directory of pathValue.split(delimiter)) {
    const candidate = join(directory, "security");
    if (existsSync(candidate)) return candidate;
  }
  return undefined;
}

function pathOptions(options: CredentialOptions): { environ?: Readonly<Record<string, string | undefined>> } {
  return options.environ === undefined ? {} : { environ: options.environ };
}
